File: backend/order/src/main/java/com/service/order/entity/Order.java
package com.service.order.entity;

import com.service.order.enums.OrderStatus;
import jakarta.persistence.*;
import jakarta.validation.constraints.Email;
import lombok.*;
import org.hibernate.annotations.DynamicUpdate;

import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@DynamicUpdate
@Table(
        name = "orders",
        indexes = {
                @Index(name = "idx_order_customer_id", columnList = "customer_id"),
                @Index(name = "idx_order_restaurant_id", columnList = "restaurant_id"),
                @Index(name = "idx_order_status", columnList = "status"),
                @Index(name = "idx_order_created_at", columnList = "created_at")
        }
)
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Order {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "customer_id", nullable = false)
    private UUID customerId;

    @Column(name = "restaurant_id", nullable = false)
    private UUID restaurantId;

    @Column(name = "payment_id")
    private UUID paymentId;

    @Column(name = "delivery_partner_id")
    private UUID deliveryPartnerId;

    @Column(name = "delivery_partner_name")
    private String deliveryPartnerName;

    @Column(name = "delivery_partner_phone", length = 20)
    private String deliveryPartnerPhone;

    @Column(nullable = false, length = 500)
    private String deliveryAddress;

    @Column(nullable = false, precision = 12, scale = 2)
    private BigDecimal subtotal;

    @Column(nullable = false, precision = 12, scale = 2)
    private BigDecimal deliveryFee;

    @Column(nullable = false, precision = 12, scale = 2)
    private BigDecimal tax;

    @Column(nullable = false, precision = 12, scale = 2)
    private BigDecimal totalAmount;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private OrderStatus status;

    @OneToMany(
            mappedBy = "order",
            cascade = CascadeType.ALL,
            orphanRemoval = true
    )
    @Builder.Default
    private List<OrderItem> items = new ArrayList<>();

    @Column(nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @Column(nullable = false)
    private OffsetDateTime updatedAt;

    @Embedded
    private DeliveryLocation deliveryLocation;

    @Column(nullable = false)
    private String customerName;

    @Column(nullable = false, length = 20)
    private String customerPhone;

    @Email
    @Column(name = "customer_email",nullable = false)
    private String customerEmail;

    @PrePersist
    public void onCreate() {
        createdAt = OffsetDateTime.now();
        updatedAt = OffsetDateTime.now();

        if (status == null) {
            status = OrderStatus.PENDING_PAYMENT;
        }
    }

    @PreUpdate
    public void onUpdate() {
        updatedAt = OffsetDateTime.now();
    }
}

File: frontend/Customer_app/lib/features/orders/domain/entities/order.dart
import 'package:decimal/decimal.dart';
import 'package:flutter/foundation.dart';

import '../order_status.dart';
import 'delivery_location.dart';
import 'order_item.dart';

/// Domain entity for a customer order.
///
/// `createdAt` is parsed from the raw DTO string into a `DateTime`.
/// Customer identity fields are nullable — they come from the server response
/// but are not sent by the client (Property 15).
@immutable
class Order {
  const Order({
    required this.id,
    this.customerId,
    this.customerName,
    this.customerPhone,
    this.customerEmail,
    this.deliveryPartnerId,
    this.deliveryPartnerName,
    this.deliveryPartnerPhone,
    required this.restaurantId,
    required this.deliveryLocation,
    required this.subtotal,
    required this.deliveryFee,
    required this.tax,
    required this.totalAmount,
    required this.status,
    required this.items,
    required this.createdAt,
  });

  final String id;
  final String? customerId;
  final String? customerName;
  final String? customerPhone;
  final String? customerEmail;
  final String? deliveryPartnerId;
  final String? deliveryPartnerName;
  final String? deliveryPartnerPhone;
  final String restaurantId;
  final DeliveryLocation deliveryLocation;
  final Decimal subtotal;
  final Decimal deliveryFee;
  final Decimal tax;
  final Decimal totalAmount;
  final OrderStatus status;
  final List<OrderItem> items;
  final DateTime createdAt;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Order && runtimeType == other.runtimeType && id == other.id;

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() =>
      'Order(id: $id, status: $status, total: $totalAmount)';
}
